## Modeling Part

In this section, we will build a machine learning model to predict the number of goals scored by both teams during a game. Predicting football match outcomes is inherently challenging due to the high degree of variability and unpredictability involved. Many matches are influenced by intangible factors such as player form, individual decisions, and situational dynamics that are difficult to capture in numerical data. Despite these complexities, our goal is to construct a model that performs as accurately as possible across a diverse range of matches.

### Challenges in Modeling Football Matches

Football match outcomes are known to be influenced by numerous variables, both observable and hidden. Even with extensive data, some matches remain inherently unpredictable due to factors such as:

- **Team Strategy and Lineups**: A change in lineup or tactical approach can significantly alter a team’s performance.
- **Player Form and Injuries**: Variations in player fitness and day-to-day form can have a substantial impact on the final outcome.
- **Psychological Factors**: Motivation, morale, and team dynamics play a significant role but are hard to quantify.

Given these nuances, building a robust predictive model requires a blend of feature engineering, advanced modeling techniques, and iterative refinement.

### Model Selection and Approach

To tackle this problem, we will employ a deep learning architecture. Neural networks are particularly well-suited for this type of task because they can capture complex interactions and relationships within large datasets. The model will consist of multiple layers that can extract and learn intricate patterns from the input features, enabling us to identify subtle trends and associations that simpler algorithms might miss.

The deep network architecture will allow us to process our extensive feature set and leverage the power of high-dimensional data representations. Specifically, we will experiment with different network configurations, such as:

- **Fully Connected Neural Networks (Dense Layers)**: Useful for capturing non-linear relationships in tabular data.
- **Dropout Layers**: To prevent overfitting by randomly deactivating certain neurons during training.
- **Activation FunctionLinearch as ReLU and Sigmoid, which help introduce non-linearity into the model and handle complex data transformations.

### Iterative Process

The process of training a reliable model will require multiple iterations and fine-tuning. We will evaluate the model's performance using various metrics and adjust hyperparameters accordingly until we achieve satisfactory results. Given the unpredictable nature of football matches, our focus will be on optimizing the model to generalize well across different scenarios, rather than just maximizing accuracy on the training set.

By leveraging the capabilities of deep learning, we aim to construct a model that can identify meaningful patterns within the data and provide robust predictions, even in a challenging and unpredictable domain like football match outcomes.


In [26]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)

import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

In [28]:
df = pd.read_csv('default_data_all_variables_v1.csv')
target = pd.read_csv('targets_v1.csv')

We need to limit the number of variables that negatively affect the model's performance, as identified during the experimentation phase. In the process of building and testing various models, we observed that certain variables introduced noise or added complexity without contributing positively to the accuracy or generalization of the model.

Although I will omit the detailed experimentation phase here, the key takeaway is that eliminating or reducing the influence of these problematic variables will improve the overall robustness of the model. This process of feature selection or reduction will help streamline the data, ensuring that the model focuses on the most impactful variables, leading to better predictions.

By carefully curating the feature set, we aim to strike a balance between having enough information to capture meaningful patterns and avoiding overfitting or unnecessary complexity. This refined approach will allow the model to generalize better across unseen data, improving its ability to predict match outcomes.


In [31]:
df = df.loc[:, ~df.columns.str.contains('HomeTeam|AwayTeam')]
df = df.loc[:, ~df.columns.str.contains('year|month')]
df = df.loc[:, ~df.columns.str.contains('ovr')]
df = df.loc[:, ~df.columns.str.contains('corner')]
df = df.loc[:, ~df.columns.str.contains('TimeBucket')]
df = df.loc[:, ~df.columns.str.contains('Avg')]

In [33]:
target = target[['FTHG', 'FTAG']]

Last glimpse of the table before we procced to the data division.

In [77]:
df

matchday  total_points  total_points_home_team  total_points_away_team  \
0      0.000000      0.000000                0.000000                0.000000   
1      0.000000      0.000000                0.000000                0.000000   
2      0.000000      0.000000                0.000000                0.000000   
3      0.000000      0.000000                0.000000                0.000000   
4      0.000000      0.000000                0.000000                0.000000   
...         ...           ...                     ...                     ...   
17252  0.088889      0.000000                0.000000                0.020619   
17253  0.088889      0.062500                0.020408                0.061856   
17254  0.088889      0.072917                0.081633                0.072165   
17255  0.088889      0.020833                0.020408                0.092784   
17256  0.088889      0.062500                0.061224                0.041237   

         5_form  5_form_home_team  5_form_away_team   10_form  \
0      0.000000          0.000000          0.000000  0.000000   
1      0.000000          0.000000          0.000000  0.000000   
2      0.000000          0.000000          0.000000  0.000000   
3      0.000000          0.000000          0.000000  0.000000   
4      0.000000          0.000000          0.000000  0.000000   
...         ...               ...               ...       ...   
17252  0.000000          0.000000          0.133333  0.000000   
17253  0.400000          0.133333          0.400000  0.200000   
17254  0.466667          0.533333          0.466667  0.233333   
17255  0.133333          0.133333          0.600000  0.066667   
17256  0.400000          0.400000          0.266667  0.200000   

       10_form_home_team  10_form_away_team  total_goals  \
0               0.000000           0.000000     0.000000   
1               0.000000           0.000000     0.000000   
2               0.000000           0.000000     0.000000   
3               0.000000           0.000000     0.000000   
4               0.000000           0.000000     0.000000   
...                  ...                ...          ...   
17252           0.000000           0.066667     0.040404   
17253           0.066667           0.200000     0.060606   
17254           0.266667           0.233333     0.080808   
17255           0.066667           0.300000     0.020202   
17256           0.200000           0.133333     0.050505   

       total_goals_home_team  total_goals_away_team  total_goals_against  \
0                   0.000000               0.000000             0.000000   
1                   0.000000               0.000000             0.000000   
2                   0.000000               0.000000             0.000000   
3                   0.000000               0.000000             0.000000   
4                   0.000000               0.000000             0.000000   
...                      ...                    ...                  ...   
17252               0.039216               0.037736             0.128713   
17253               0.039216               0.056604             0.049505   
17254               0.049020               0.075472             0.059406   
17255               0.019608               0.066038             0.069307   
17256               0.049020               0.056604             0.049505   

       total_goals_against_home_team  total_goals_against_away_team  \
0                           0.000000                           0.00   
1                           0.000000                           0.00   
2                           0.000000                           0.00   
3                           0.000000                           0.00   
4                           0.000000                           0.00   
...                              ...                            ...   
17252                       0.128713                           0.07   
17253                       0.089109               

Now, we need to divide the dataset into training, validation, and test sets. This is a critical step in the modeling process to ensure that our model is trained effectively and evaluated properly. In this case, the split is done manually, with a specific strategy in mind.

### Data Splitting Strategy

Instead of using a random split, I opted for a chronological split of the data. The reasoning behind this approach is to leave the most recent observations as the test set, mimicking real-world conditions where we use the model to predict upcoming matches. This simulates the process of applying the model to live betting scenarios, where the latest games are the ones we aim to predict.

- **Training Set**: This portion of the data contains the bulk of historical matches and is used to train the model. By training on a wide range of past games, we allow the model to learn from various situations and patterns.
  
- **Validation Set**: A smaller, intermediate set that allows us to tune hyperparameters and evaluate the model’s performance during the training process. This helps us avoid overfitting by ensuring that the model generalizes well to unseen data during training.
  
- **Test Set**: The most recent matches are set aside as the final test set. These observations will be used as a final evaluation of the model before deploying it in a real betting scenario. By doing this, we ensure that the test set reflects the same conditions the model will face when making predictions in live matches.

This manual approach to splitting the data ensures that we are working with a realistic time-series setup, where past data is used to predict future outcomes. It also prepares the model for its ultimate goal: performing well on unseen matches.


In [35]:
X_train = df[:-1000]
y_train = target[:-1000]
X_val = df[-1000:-550]
y_val = target[-1000:-550]
X_test = df[-550:-40]
y_test = target[-550:-40]
X_to_pred = df[-40:]

Now, we will construct the Deep Neural Network (DNN) architecture for our model. The primary objective is to predict the number of goals scored by both teams, so we need to design a neural network that can effectively learn from the input data and generalize well to new, unseen matches.

### DNN Architecture

1. **Input and Output Layers**: 
   We begin by defining the input and output layers, ensuring that their sizes are compatible with our training data and target data. The input layer will have a size equal to the number of features in our dataset, while the output layer will have a size of 2 (for the goals scored by the home and away teams).

2. **Hidden Layers**:
   To enhance the learning capability of the model, we add two hidden layers. These hidden layers enable the network to capture more complex patterns and interactions in the data. A typical approach is to use fully connected (dense) layers, where each neuron is connected to all neurons in the previous layer.
   
   The activation function used in these layers will be **ReLU (Rectified Linear Unit)**, which introduces non-linearity to the model and helps it learn complex relationships between the features.

3. **Dropout Layer**:
   To prevent overfitting, we include a **dropout layer**, which randomly deactivates a fraction of neurons during training. This regularization technique encourages the model to be more general and not rely too heavily on specific neurons, improving its robustness.

4. **Training Hyperparameters**:
   We will manually set the learning rate, batch size, and number of epochs to control the training process:
   - **Learning Rate**: Controls how quickly the model updates its weights during training.
   - **Batch Size**: The number of training examples processed in one iteration before updating the model weights.
   - **Epochs**: The number of times the entire training dataset is passed
     through the model.

5. **Early Stopping**:
   To avoid overfitting, we will implement an **early stopping** mechanism. This technique monitors the model’s performance on the validation set and stops training when the model's performance stops improving. This ensures that the model doesn’t continue to train unnecessarily, which could lead to overfitting.

### Evaluation Metrics

We will evaluate the model using two metrics:
- **Mean Squared Error (MSE)**: This metric measures the average squared difference between the predicted and actual values. While useful, it can be sensitive to outliers, which is why we focus primarily on the next metric.
- **Mean Absolute Error (MAE)**: MAE gives us the mean absolute difference between the predicted and actual goals, making it more interpretable for our specific problem. This metric directly shows how far off our predictions are from the actual number of goals scored, which is crucial for real-world betting applications.

By optimizing the model based on MAE, we will be able to better assess the average error in our predictions for goals scored, providing more meaningful insights into the model’s accuracy.


In [ ]:
from tensorflow.keras.optimizers import Adam, RMSprop, SGD
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.regularizers import l2, l1
import random
random.seed(42)
np.random.seed(42)
tf.random.set_seed(42)

optimizer = Adam(learning_rate=0.01)
model = Sequential()
# model.add(BatchNormalization())
model.add(Dense(256, activation='relu', input_shape=(199,)))  # 278 is the number of input features
model.add(Dropout(0.1)) 
model.add(Dense(64, activation='relu', kernel_regularizer=l1(0.001)))
# model.add(Dropout(0.1))  # Optional, to prevent overfitting
model.add(Dense(16, activation='relu'))
model.add(Dense(2, activation='linear'))
model.compile(optimizer=optimizer, loss='mse', metrics=['mae'])
early_stopping = EarlyStopping(monitor='val_loss', patience=30)
history = model.fit(X_train, y_train, epochs=100, batch_size=500, validation_data=(X_val, y_val), callbacks=[early_stopping])
results = model.evaluate(X_test, y_test)


For our initial deep learning model, we obtained a mean **Mean Absolute Error (MAE)** of about 1.2 on the test set. While this result is a good starting point, it leaves room for improvement. To enhance the model's performance, we will incorporate a **hyperparameter optimization tool** provided by Keras. This tool will allow us to fine-tune various aspects of the model by searching for the optimal values of key hyperparameters.

### Hyperparameter Optimization with Keras

Keras provides a powerful tool for hyperparameter tuning, which can automatically search through a range of possible values for different parameters. This optimization process is essential because the performance of a neural network is heavily influenced by the specific configuration of its hyperparameters.

Instead of manually adjusting the parameters through trial and error, we will use a **random search algorithm** to systematically explore different combinations of hyperparameters. By leveraging this tool, we aim to significantly improve the model’s performance, especially in terms of reducing the prediction error (MAE).

### Hyperparameters to Tune

The following hyperparameters will be optimized during the random search process:
1. **Number of Hidden Layers**: The number of hidden layers in the network plays a crucial role in determining the depth of the model and its ability to capture complex patterns. The optimal number of hidden layers will be determined through the search process.
  
2. **Number of Perceptrons (Neurons) in Each Layer**: For each hidden layer, the number of perceptrons (neurons) can vary. Too few neurons may lead to underfitting, while too many neurons may result in overfitting. Finding the right balance is key to improving the model's generalization.

3. **Learning Rate**: The learning rate controls the step size during the model’s optimization process. A learning rate that is too high can cause the model to converge too quickly and miss optimal solutions, while a rate that is too low can lead to slow training. We will search for the optimal learning rate to speed up convergence without sacrificing accuracy.

4. **Dropout Rate**: Dropout is a regularization technique used to prevent overfitting by randomly deactivating a percentage of neurons during training. We will tune the dropout rate to find the best balance between model complexity and generalization.

5. **Activation Function**: Different activation functions can be used in hidden layers to introduce non-linearity. The choice of activation function (e.g., ReLU, Sigmoid, Tanh) can significantly impact how well the model learns complex relationships in the data. The random search will explore various options.

6. **Number of Epochs**: The number of epochs determines how many times the model will pass through the entire training dataset. Too few epochs might result in underfitting, while too many might lead to overfitting. We will tune this parameter to find the optimal number of epochs.

7. **Batch Size**: The batch size specifies how many samples are processed before updating the model’s weights. Smaller batch sizes provide more frequent updates, while larger ones can speed up training but may miss finer details. Tuning this parameter will help improve model efficiency and performance.

### Random Search Algorithm

The random search algorithm will iteratively try different combinations of the hyperparameters listed above, and evaluate each combination based on the **validation set** performance (e.g., using MAE as the metric). This process will help us identify the most effective model configuration without the need for exhaustive grid searches, which can be computationally expensive.

### Goal

By utilizing hyperparameter optimization, we aim to improve the model’s mean MAE on the test set, bringing it below the current value of 1.2. Optimizing the model in this way will allow us to extract more meaningful patterns from the data and make more accurate predictions, which is especially important when forecasting unpredictable events like football matches.


In [1659]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, BatchNormalization
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.regularizers import l2
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.optimizers.schedules import ExponentialDecay
from kerastuner import HyperModel
from kerastuner.tuners import RandomSearch

# Define the model in a HyperModel class
class RegressionHyperModel(HyperModel):

    def build(self, hp):
        model = Sequential()

        # Input layer
        model.add(Dense(hp.Int(f'units_1', min_value=250, max_value=600, step=10), activation='relu', input_shape=(199,)))
        model.add(Dropout(hp.Float(f'dropout_1', min_value=0.0, max_value=0.5, step=0.1)))
        model.add(Dense(hp.Int(f'units_2', min_value=48, max_value=256, step=5), activation='relu',))
   
        model.add(Dense(hp.Int(f'units_3', min_value=8, max_value=128, step=4), activation='relu',))
        # Output layer
        model.add(Dense(2, activation='linear'))

        # Learning Rate Decay with Exponential Decay
        initial_learning_rate = hp.Float('learning_rate', min_value=1e-4, max_value=5*1e-2, sampling='LOG')

        # Optimizer with learning rate decay
        optimizer = Adam(learning_rate=initial_learning_rate)
        early_stopping = EarlyStopping(monitor='val_loss', patience=30)
        # Compile model
        model.compile(optimizer=optimizer, loss='mse', metrics=['mae'])
        
        return model

# Instantiate the HyperModel
hypermodel = RegressionHyperModel()

# Create the tuner
tuner = RandomSearch(
    hypermodel,
    objective='val_mae',  # Objective is to minimize validation MAE
    max_trials=500,        # Number of different hyperparameter combinations to try
    executions_per_trial=1,  # Number of executions for each trial
    directory='hyperparameter_search3',  # Where to save the results
    project_name='regression_tuning3'
)

# Early stopping to prevent overfitting
early_stopping = EarlyStopping(monitor='val_loss', patience=30)

# Run hyperparameter search
tuner.search(X_train, y_train, 
             epochs=500,  # Start with a static number of epochs, can adjust if needed
             batch_size=9000,  # You can also tune this if desired
             validation_data=(X_val, y_val), 
             callbacks=[early_stopping])

# Retrieve the best hyperparameters
best_hps = tuner.get_best_hyperparameters(num_trials=1)[0]

# Train the best model with optimal hyperparameters including batch size and epochs
history = tuner.hypermodel.build(best_hps).fit(
    X_train, y_train,
    epochs=best_hps.get('epochs') if 'epochs' in best_hps.values else 1000,  # Add a fallback for epochs
    batch_size=best_hps.get('batch_size') if 'batch_size' in best_hps.values else 9000,  # Add a fallback for batch size
    validation_data=(X_val, y_val),
    callbacks=[early_stopping]
)

# Evaluate the model on test data
results = tuner.hypermodel.build(best_hps).evaluate(X_test, y_test)



Reloading Tuner from hyperparameter_search3\regression_tuning3\tuner0.json
Epoch 1/1000


C:\Users\ciesl\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


2/2 ━━━━━━━━━━━━━━━━━━━━ 1s 195ms/step - loss: 8.0943 - mae: 2.1976 - val_loss: 2.9380 - val_mae: 1.2781
Epoch 2/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step - loss: 2.7098 - mae: 1.2351 - val_loss: 1.5796 - val_mae: 0.9851
Epoch 3/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step - loss: 1.6801 - mae: 1.0029 - val_loss: 2.3395 - val_mae: 1.3085
Epoch 4/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step - loss: 2.0078 - mae: 1.1653 - val_loss: 1.7934 - val_mae: 1.1306
Epoch 5/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step - loss: 1.6364 - mae: 1.0315 - val_loss: 1.4737 - val_mae: 0.9476
Epoch 6/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 48ms/step - loss: 1.5731 - mae: 0.9558 - val_loss: 1.5179 - val_mae: 0.9498
Epoch 7/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 38ms/step - loss: 1.6344 - mae: 0.9637 - val_loss: 1.5110 - val_mae: 0.9606
Epoch 8/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step - loss: 1.5711 - mae: 0.9574 - val_loss: 1.4488 - val_mae: 0.9670
Epoch 9/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 38ms/step - loss: 1.4874 - mae: 0.9

C:\Users\ciesl\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 3.6198 - mae: 1.4427  


After identifying the best-performing combination of hyperparameters through the random search process, we can integrate these optimized values into our primary deep learning architecture. By doing so, we ensure that the model is configured to make the most accurate predictions based on the specific characteristics of our dataset.

### Applying Optimized Hyperparameters

Once we have the optimal hyperparameters (such as the number of hidden layers, number of perceptrons, learning rate, dropout rate, activation function, number of epochs, and batch size), we will modify our original model’s architecture to incorporate these values. This process will fine-tune the network for better performance and more accurate predictions.

### Rerunning the Model

With the optimized architecture in place, we will retrain the model on the training set and validate its performance on the test set. By doing this, we can directly observe the improvements resulting from the hyperparameter optimization.

1. **Training the Optimized Model**: The model will be trained again, but now with the hyperparameters that were found to yield the best results during the optimization phase. We expect the model to converge more effectively, leading to improved predictive performance.

2. **Evaluating on the Test Set**: After training, we will evaluate the model on the test set to observe how well it generalizes to unseen data. The test set provides a realistic evaluation of the model’s ability to predict the number of goals scored by both teams.

### Expected Results

By using the optimized hyperparameters, we expect a reduction in the model's **Mean Absolute Error (MAE)** on the test set, compared to the initial architecture. A lower MAE indicates that the model has improved in terms of predicting the actual goals scored during a match, which is critical for making more accurate football match predictions.

We will closely analyze the test set results to ensure the improvements are significant and consistent across various matches. If necessary, further fine-tuning can be performed, but with the optimized architecture, we should see a noticeable enhancement in the model's overall performance.


In [31]:
from tensorflow.keras.optimizers import Adam, RMSprop, SGD
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.regularizers import l2, l1
import random
random.seed(42)
np.random.seed(42)
tf.random.set_seed(42)

optimizer = Adam(learning_rate=0.013556832247938065)
model = Sequential()
# model.add(BatchNormalization())
model.add(Dense(320, activation='relu', input_shape=(199,)))  # 278 is the number of input features
model.add(Dropout(0.2)) 
model.add(Dense(83, activation='relu', kernel_regularizer=l1(0.001)))
# model.add(Dropout(0.1))  # Optional, to prevent overfitting
model.add(Dense(12, activation='relu'))
model.add(Dense(2, activation='linear'))
model.compile(optimizer=optimizer, loss='mse', metrics=['mae'])
early_stopping = EarlyStopping(monitor='val_loss', patience=30)
history = model.fit(X_train, y_train, epochs=1000, batch_size=9000, validation_data=(X_val, y_val), callbacks=[early_stopping])
results = model.evaluate(X_test, y_test)


Epoch 1/1000


C:\Users\ciesl\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


2/2 ━━━━━━━━━━━━━━━━━━━━ 2s 186ms/step - loss: 7.8101 - mae: 1.7792 - val_loss: 8.2953 - val_mae: 2.3686
Epoch 2/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 34ms/step - loss: 6.2054 - mae: 1.8579 - val_loss: 4.1479 - val_mae: 1.3044
Epoch 3/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 38ms/step - loss: 4.1294 - mae: 1.3198 - val_loss: 3.9703 - val_mae: 1.2910
Epoch 4/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 35ms/step - loss: 3.7904 - mae: 1.2627 - val_loss: 2.8882 - val_mae: 1.0486
Epoch 5/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step - loss: 2.8655 - mae: 1.0449 - val_loss: 2.9919 - val_mae: 1.2227
Epoch 6/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 35ms/step - loss: 2.8881 - mae: 1.1458 - val_loss: 2.8935 - val_mae: 1.1892
Epoch 7/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step - loss: 2.6650 - mae: 1.0698 - val_loss: 2.3651 - val_mae: 0.9757
Epoch 8/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 33ms/step - loss: 2.4845 - mae: 0.9663 - val_loss: 2.4306 - val_mae: 0.9604
Epoch 9/1000
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 33ms/step - loss: 2.5731 - mae: 0.9

In [32]:
results = model.evaluate(X_test, y_test)
results

16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 1.4863 - mae: 0.9542 


[1.431609869003296, 0.9289669990539551]

After rerunning the model with the optimized hyperparameters, we can observe a significant reduction in the prediction error. The **Mean Absolute Error (MAE)** has declined substantially, reaching a value of around **0.92** on the test set. This result is a clear indication that the hyperparameter tuning process has greatly improved the model’s performance.

### Conclusion on Model Performance

The reduction in MAE demonstrates that the optimized architecture is well-suited for this type of football match prediction task. Achieving an MAE of around **0.92** suggests that, on average, the model’s predictions are quite close to the actual goals scored by both teams. This level of accuracy is a considerable improvement from our initial attempts and represents a strong result for such a complex and unpredictable problem.

### Challenges and Random Factors

While this value of **0.92** is close to optimal for our model, it is important to recognize the inherent challenges of this task:
- Football matches are influenced by a variety of random factors such as team morale, unexpected events (like injuries or red cards), and form fluctuations that are difficult to capture in data-driven models.
- Human elements, such as players' physical condition and managerial decisions, add unpredictability to the game, making it extremely challenging to obtain more accurate predictions through modeling alone.

### Final Thoughts

Given the complexity of the task and the various random factors affecting match outcomes, further reducing the MAE beyond **0.92** would be immensely difficult without introducing additional, high-quality data or advanced techniques. However, the current results demonstrate that our model has achieved a reliable level of predictive accuracy, making it a strong tool for predicting goals in football matches.

We can now confidently proceed with this model, knowing that it is well-tuned for the task at hand, and further attempts to improve it may yield diminishing returns due to the random nature of the domain. Let's save the model for further usage.

In [37]:
# model.save("model_fully_optimal.keras")
from tensorflow.keras.models import load_model
# # 
model = load_model("model_fully_optimal.keras")
model.evaluate(X_test, y_test)

16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 1.5500 - mae: 0.9698  


[1.4403669834136963, 0.9361046552658081]

### Matching Team Names and Over/Under 2.5 Goals Odds to Fixtures

For the test dataset, the next step involves matching the team names and over/under 2.5 goals odds to the corresponding fixtures. This is crucial because it allows us to align our predictions with the actual match data, so we know exactly what we are dealing with for each game.

To do this, we will:

1. **Match Team Names**: 
   First, we will ensure that the team names in the test set align with those in the training set and the original datasets. This step is important to avoid discrepancies in naming conventions (e.g., abbreviations, misspellings). Any necessary transformations or mappings will be applied here.

2. **Link Over/Under 2.5 Goals Odds**:
   The next task is to incorporate the over/under 2.5 goals odds into the test dataset. These odds are key to determining whether the bookmaker expects a high-scoring game (over 2.5 goals) or a low-scoring one (under 2.5 goals). By merging the odds with the test data, we will be able to directly compare our goal predictions with the bookmaker’s odds.

### Summing Predicted Goals and Comparing with Actual Results

Once we have aligned the team names and merged the odds, we can proceed to calculate the predicted total goals for each match and compare it with the actual results.

1. **Summing Predicted Goals**:
   After obtaining the predictions for goals scored by both teams, we will sum the predicted goals for each match to get a total goal prediction. This total will be used to determine whether the predicted number of goals is higher or lower than 2.5, matching the over/under 2.5 goal betting line.

2. **Comparison with Actual Results**:
   For each fixture in the test set, we will then compare:
   - The **predicted total goals** with the **actual total goals** scored in the match.
   - The bookmaker's **over/under 2.5 odds** with our prediction of whether the total goals will be over or under 2.5.

This comparison allows us to evaluate the performance of our model in terms of correctly predicting the total goals scored and how well our predictions align wiareas for potential improvement.


Additionally, we need to find a way to make the model useful in gambling and figure out how to incorporate it into betting on matches so that it generates profit. This involves developing a strategy where the model's predictions can be compared with bookmaker odds to identify profitable opportunities. By leveraging the model’s predictions, we aim to place smart bets, maximize returns, and manage risks, ultimately using it to gain an edge in the betting market.


In [39]:
y_pred = model.predict(X_test)

res = pd.DataFrame(y_pred, columns = ['FTHG', 'FTAG'])

over_under2 = pd.read_csv('2_5_goals.csv')[-550:-40]
over_under2.reset_index(inplace = True, drop = True)

trg = target[-550:-40]
trg.reset_index(inplace = True, drop = True)
next_fixtures2 = pd.read_csv('next_fixture_teams.csv')[-550:-40]
next_fixtures2.reset_index(inplace = True, drop = True)
over_under2 = pd.read_csv('2_5_goals.csv')[-700:-40]
over_under2.reset_index(inplace = True, drop = True)
next_fixtures3 = pd.merge(next_fixtures2[['HomeTeam',	'AwayTeam']], over_under2, on =['HomeTeam',	'AwayTeam'], how = 'left')
next_fixtures3.drop_duplicates(['HomeTeam',	'AwayTeam'],inplace= True, keep = 'last')

next_fixtures2[['FTHG', 'FTAG']] = y_pred
next_fixtures2['goals_total_pred'] = next_fixtures2['FTHG']+next_fixtures2['FTAG']
next_fixtures2[['goals_home', 'goals_away']] = trg[['FTHG', 'FTAG']]
next_fixtures2['total_goals_scored'] = next_fixtures2['goals_home']+next_fixtures2['goals_away']

next_fixtures4 = pd.merge(next_fixtures2, next_fixtures3, on =['HomeTeam',	'AwayTeam'], how = 'left')


16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


In [41]:
next_fixtures4.head()

HomeTeam    AwayTeam        Date      FTHG      FTAG  goals_total_pred  \
0   Mirandes  Valladolid  2024-05-04  1.125890  1.326753          2.452643   
1     Huesca      Oviedo  2024-05-04  1.228806  1.100068          2.328874   
2  Santander       Elche  2024-05-04  1.354604  1.418217          2.772822   
3  Cartagena    Alcorcon  2024-05-04  1.279109  0.873631          2.152740   
4   Dortmund    Augsburg  2024-05-04  1.605459  0.939871          2.545331   

   goals_home  goals_away  total_goals_scored  Avg>2.5  Avg<2.5  
0         0.0         1.0                 1.0     2.45     1.52  
1         0.0         2.0                 2.0     2.84     1.40  
2         3.0         1.0                 4.0     2.01     1.77  
3         1.0         0.0                 1.0     2.66     1.45  
4         5.0         1.0                 6.0     1.32     3.33

### Betting Strategy: Over/Under 2.5 Goals Line

The primary strategy we are implementing revolves around using the **over/under 2.5 goals betting line** to estimate how much profit we could generate by betting on either side of the market before each match. In this approach, we aim to test whether consistently betting on **over 2.5 goals** or **under 2.5 goals** yields a profit over time.

#### Assumptions and Betting Mechanics

For each bet, we assume a **fixed stake of 1 unit**. The calculation of potential winnings is as follows:

1. **Over 2.5 Goals Bet**:
   - If a match ends with **more than 2.5 goals** (i.e., 3 or more goals are scored), we win the bet.
   - Our winnings for a successful bet on over 2.5 goals are calculated using the following formula:

$ \text{Winnings} = 1 \times \text{Over 2.5 odds} \times 0.88$

Here, the **over 2.5 odds** represent the bookmaker's odds for the bet, and the factor of **0.88** accounts for tax deductions (assuming an 12% tax on winnings). If our bet wins, this formula gives us the payout based on the odds provided by the bookmaker.

2. **Under 2.5 Goals Bet**:
   - Conversely, if we decide to bet on **under 2.5 goals**, and the match ends with **fewer than 2.5 goals** (i.e., 0, 1, or 2 goals are scored), the same formula applies:

$\text{Winnings} = 1 \times \text{Under 2.5 odds} \times 0.88 $


Just like with the over 2.5 goals bet, this formula calculates the payout if our bet on under 2.5 goals is successful.

3. **No Payout for Losing Bets**:
   - In cases where the result of the match does not align with our bet, the payout is 0. For example, if we bet on over 2.5 goals and the match ends with fewer than 2.5 goals, or we bet on under 2.5 goals and more than 2.5 goals are scored, the payout is:
   

$\text{Winnings} = 0$

This simple win/loss dynamic ensures that our betting approach is straightforward, and it allows us to evaluate how often the model's predictions align with the actual match outcomes.

In [85]:
def money_won_goals_over(row, goals):
    if row['goals_total_pred'] > goals:
        if row['total_goals_scored'] >2.5:
             return row['Avg>2.5']
        else:
            return 0
    else:
        return 0
    
def money_won_goals_under(row, goals):
    if row['goals_total_pred'] < goals:
        if row['total_goals_scored'] <2.5:
             return row['Avg<2.5']
        else:
            return 0
    else:
        return 0

### Strategy to Profit from the Model: Over 2.5 Goals Betting Example

The way we aim to profit from this model is somewhat complex, as it involves multiple layers of decision-making and parameter tuning. Let’s break down the process, using **betting on over 2.5 goals** as an example.

#### Step 1: Setting Thresholds for Model Predictions and Odds

To implement this strategy, we first need to propose two threshold values:

1. **Threshold for predicted goals (t1):** This value is based on the model’s predictions for the number of goals scored in a match. The range of possible values for this threshold is approximately **t1 ~ [0.5, 4.5]**. This means we only place a bet on over 2.5 goals if the predicted goals by the model for a match exceed this threshold.
   
2. **Threshold for over 2.5 odds (t2):** This is the minimum value for the bookmaker’s odds for betting on over 2.5 goals. The range for this threshold is **t2 ~ [1.4, 2.6]**, meaning we only bet if the bookmaker’s odds for over 2.5 goals exceed this threshold.

#### Step 2: Filtering Matches Based on Thresholds

For each combination of these thresholds, **t1** and **t2**, we filter the matches to only include those where both conditions are met:
- The model predicts **more than t1 goals** for the match.
- The bookmaker’s odds for **over 2.5 goals** are greater than **t2**.

By varying the thresholds, we create different groups of matches. Each group represents a set of matches where both the predicted goals and the odds exceed the respective thresholds. These groups will form the basis for our betting strategy.

#### Step 3: Calculating Yield and Money Won

For each group of matches that meets the conditions (i.e., predicted goals > t1 and odds > t2), we calculate two key metrics:
1. **Yield:** This is a common measure in gambling used to assess profitability. The **yield** is calculated as:

$\text{Yield} = \frac{\text{Total Money Won from All Matches}}{\text{Total Stake (Number of Bets)}}$

 In simpler terms, the yield represents the average return on each bet placed in that group of matches. A yield greater than 1 indicates that the strategy is profitable, while a yield less than 1 suggests a loss.

   - **Explanation of Yield in Gambling:** The yield is a key performance indicator in gambling. It reflects how much profit or loss is made relative to the total amount wagered. A yield of 1 would mean breaking even (i.e., no profit or loss). For example, a yield of 1.10 would mean a 10% profit on the total amount wagered, while a yield of 0.90 would indicate a 10% loss.

2. **Total Money Won:** The amount of money won for each group of matches is calculated using the formula:

$\text{Total Money Won} = (\text{Yield} - 1) \times \text{Number of Matches in the Group}$



   This calculation gives us the overall profit or loss from betting on the matches that meet the threshold conditions for **t1** and **t2**. The total stake in each case is the number of matches in that group (since we assume a stake of 1 unit per match).

#### Example of Calculation:

Let’s assume we have the following thresholds:
- t1 = 2.8 (the model predicts more than 2.8 goals)
- t2 = 1.8 (the bookmaker’s odds for over 2.5 goals are greater than 1.8)

We filter the matches to include only those where the model’s predicted goals exceed 2.8 and the odds for over 2.5 goals are greater than 1.8. For this group, we calculate the total money won and yield.

If, after evaluating all matches that meet these criteria, the yield is 1.15, this means that on average, we make a 15% profit on each bet placed in this group. If 100 matches meet the conditions, the total money won would be:

$\text{Total Money Won} = (1.15 - 1) \times 100 = 0.15 \times 100 = 15 \text{ units of profit}$

#### Step 4: Testing and Optimizing Thresholds

To maximize profitability, we will test different combinations of **t1** and **t2** across a range of values. By analyzing the yield and total money won for each combination, we aim to identify the optimal thresholds that maximize long-term profit. This process allows us to fine-tune the model to generate consistent returns from betting on over 2.5 goals.

### Summary

The overall strategy involves filtering matches based on thresholds for predicted goals and bookmaker odds, and then calculating the yield and profit for each group of matches that meets the criteria. By testing various threshold combinations, we can refine our betting strategy to achieve the highest possible yield, thereby increasing profitability in the long run.


In [87]:
range_1 = np.linspace(0.5,4.5,81)
range_2 = np.linspace(1.4,2.6,61)

In [89]:
matches_list = []
winnings_list = []
yield_list = []
i_s = []
j_s = []
money_won_over_all_matches = []
for i in range_1:
    for j in range_2:
        next_fixtures4['stake'] = next_fixtures4.apply(money_won_goals_over, axis=1,args=(i,))
        no_of_matches = len(next_fixtures4.loc[(next_fixtures4['goals_total_pred'] > i) & (next_fixtures4['Avg>2.5']>j)])
        winnings = next_fixtures4.loc[(next_fixtures4['goals_total_pred'] > i) & (next_fixtures4['Avg>2.5']>j)]['stake'].sum()*0.88
        if no_of_matches != 0:
            yield_ratio = winnings / no_of_matches
        else:
            yield_ratio = 0
        matches_list.append(no_of_matches)
        winnings_list.append(winnings)
        yield_list.append(yield_ratio)
        i_s.append(i)
        j_s.append(j)
        money_won_over_all_matches.append((yield_ratio-1)*no_of_matches)
results_df_over = pd.DataFrame({'goals_total_over':i_s,
                                'over_2_5_val': j_s,
                                'number_of_matches': matches_list,
                                'money_won': winnings_list,
                                'yield':yield_list,
                                'money_won_over_all_matches':money_won_over_all_matches
                               })

In [90]:
over_conditions = results_df_over.sort_values('money_won_over_all_matches', ascending = False).head(15)

### Practical Example of Calculations

In reality, after running our model and applying the thresholds, we can organize the results into a table. The table lists various combinations of predicted goals (t1) and over 2.5 odds (t2), and is **sorted by the highest amount of money won for each group** of matches. The key metrics displayed in this table include the **number of matches**, the **yield**, and the **total money won** for each group that meets the respective threshold conditions.

#### Observations:

As we analyze the table, we can notice several interesting patterns:
- **Substantial Number of Groups with Yield > 1**: A considerable number of match groups demonstrate a **yield greater than 1**, indicating that these groups are profitable. A yield higher than 1 means that, on average, the bets placed on these matches resulted in a positive return. This suggests that the model is successfully identifying profitable opportunities based on the combination of predicted goals and the offered odds.
  
- **Significant Total Money Won**: In addition to the positive yield, many of these groups show a **total money won** that exceeds **2 units**. This means that, for those specific combinations of thresholds, the model's predictions not only resulted in positive returns but also generated a meaningful amount of profit.

#### Example:

For instance, consider one group where:
- The **predicted goals threshold (t1)** is set at 2.7, meaning we only bet on matches where the model predicts more than 2.7 goals.
- The **over 2.5 odds threshold (t2)** is set at 2.0, meaning we only place a bet if the odds for over 2.5 goals exceed 2.0.

If this group contains 28 matches, with a yield of 1.115 and total money won equal to 3.24 units, this would imply:
- A **11.5% average profit** on the amount staked across all 28 matches.
- The **total profit** for this group of matches is **3.24 units**, meaning we made 3.24 units of profit from betting on these matches with the given thresholds.

In [91]:
results_df_over.sort_values('money_won_over_all_matches', ascending = False).head(15)

goals_total_over  over_2_5_val  number_of_matches  money_won     yield  \
3311              3.20          1.74                 13    16.5440  1.272615   
2714              2.70          2.00                 28    31.2400  1.115714   
2653              2.65          2.00                 34    37.1096  1.091459   
3312              3.20          1.76                 12    15.0040  1.250333   
2656              2.65          2.06                 27    29.9464  1.109126   
3371              3.25          1.72                 12    14.8808  1.240067   
3372              3.25          1.74                 12    14.8808  1.240067   
3369              3.25          1.68                 18    20.8736  1.159644   
3310              3.20          1.72                 14    16.5440  1.181714   
3308              3.20          1.68                 20    22.5368  1.126840   
3313              3.20          1.78                 11    13.4464  1.222400   
3317              3.20          1.86                 11    13.4464  1.222400   
3314              3.20          1.80                 11    13.4464  1.222400   
3315              3.20          1.82                 11    13.4464  1.222400   
3316              3.20          1.84                 11    13.4464  1.222400   

      money_won_over_all_matches  
3311                      3.5440  
2714                      3.2400  
2653                      3.1096  
3312                      3.0040  
2656                      2.9464  
3371                      2.8808  
3372                      2.8808  
3369                      2.8736  
3310                      2.5440  
3308                      2.5368  
3313                      2.4464  
3317                      2.4464  
3314                      2.4464  
3315                      2.4464  
3316                      2.4464

In [1111]:
over_conditions.to_csv('over_conditions.csv', index= False)

### Maximizing Matches and Profits

To further enhance the betting strategy, we can increase the number of matches available for betting by focusing on the **15 most profitable groups** identified from our analysis. These groups represent the best combinations of predicted goals (t1) and over 2.5 odds (t2), which generated the highest yields and total profits.

#### Overall Profitability

When we combine the results from these 15 most profitable groups, the **overall profitability** comes out to be **4.83** units. This means that, across all the matches considered in these groups, if we had placed a **1-unit stake on each match**, by the end of the analyzed period, we would have accumulated a profit of **4.83 units**.

In other words:
- **Total number of bets placed:** One unit per match for each match in the top 15 groups.
- **Overall profit:** After summing up the wins and losses from all these bets, we are left with 4.83 units of profit.

This suggests that the combination of thresholds for these 15 groups has proven to be particularly effective, yielding consistent profits across a larger set of matches.

#### Why This Matters:

Maximizing the number of matches available to bet on, while still maintaining profitability, is crucial for generating sustained returns. By focusing on the top 15 groups:
- We ensure that our model is not overly restrictive, allowing us to place bets on a substantial number of matches.
- At the same time, we maintain a profitable betting strategy, as shown by the 4.83-unit gain, ensuring that the overall approach remains successful in the long term.

This balance between maximizing the number of opportunities and ensuring profitability is key to a robust and scalable betting strategy.


In [92]:
all_considered_matches_over = pd.DataFrame(columns = next_fixtures4.columns)
for index, row in over_conditions.iterrows():
    next_fixtures4['stake'] = next_fixtures4.apply(money_won_goals_over, axis=1,args=(row['goals_total_over'],))
    matches_to_add = next_fixtures4.loc[(next_fixtures4['goals_total_pred']>row['goals_total_over']) & (next_fixtures4['Avg>2.5']>row['over_2_5_val'])]
    all_considered_matches_over = pd.concat([all_considered_matches_over, matches_to_add])
all_considered_matches_over.drop_duplicates(inplace= True)

C:\Users\ciesl\AppData\Local\Temp\ipykernel_36744\1991751933.py:5: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  all_considered_matches_over = pd.concat([all_considered_matches_over, matches_to_add])


In [93]:
(all_considered_matches_over['stake'].sum()*0.88/len(all_considered_matches_over)-1)*len(all_considered_matches_over)

4.833599999999997

### Betting on Under 2.5 Goals

The same reasoning applies when it comes to betting on **under 2.5 goals** in matches. In this case, while there are **fewer matches** that meet the conditions for betting, the **accumulated yield** is significantly higher, making this strategy highly profitable.

#### Higher Yield with Fewer Matches

Although the number of potential matches to bet on is smaller compared to over 2.5 goals, the return on investment for these bets is more substantial. After analyzing the top groups for under 2.5 goals betting, we observe that the **total yield** across all matches leads to a profit of over **11 units**.

This means:
- **Fewer betting opportunities:** Since the under 2.5 goals line typically applies to lower-scoring games, the number of matches where the odds and predicted goals align with our thresholds is naturally reduced.
- **Higher profitability per bet:** Despite the smaller number of matches, the higher yield indicates that these bets are more reliable and profitable on average.

#### Total Profit:

With this strategy, if we had placed a **1-unit stake on each match** in the qualifying groups, the **total profit** across all matches would be **11 units**. This demonstrates that focusing on under 2.5 goals betting, even with fewer matches, can result in a more **lucrative outcome** compared to over 2.5 goer 2.5 goals markets.


In [95]:
matches_list_under = []
winnings_list_under = []
yield_list_under = []
i_s_under = []
j_s_under = []
money_won_over_all_matches_under = []
for i in range_1:
    for j in range_2:
        next_fixtures4['stake'] = next_fixtures4.apply(money_won_goals_under, axis=1,args=(i,))
        no_of_matches_under = len(next_fixtures4.loc[(next_fixtures4['goals_total_pred'] < i) & (next_fixtures4['Avg<2.5']>j)])
        winnings_under = next_fixtures4.loc[(next_fixtures4['goals_total_pred'] < i) & (next_fixtures4['Avg<2.5']>j)]['stake'].sum()*0.88
        if no_of_matches_under != 0:
            yield_ratio_under = winnings_under / no_of_matches_under
        else:
            yield_ratio_under = 0
        matches_list_under.append(no_of_matches_under)
        winnings_list_under.append(winnings_under)
        yield_list_under.append(yield_ratio_under)
        i_s_under.append(i)
        j_s_under.append(j)
        money_won_over_all_matches_under.append((yield_ratio_under-1)*no_of_matches_under)
results_df_under = pd.DataFrame({'goals_total_under':i_s_under,
                                'under_2_5_val': j_s_under,
                                'number_of_matches': matches_list_under,
                                'money_won': winnings_list_under,
                                'yield':yield_list_under,
                                'money_won_over_all_matches':money_won_over_all_matches_under
                               })

In [96]:
under_conditions = results_df_under.sort_values('money_won_over_all_matches', ascending = False).head(10)

In [97]:
results_df_under.sort_values('money_won_over_all_matches', ascending = False).head(10)

goals_total_under  under_2_5_val  number_of_matches  money_won  \
2608               2.60           2.32                 35    46.7808   
2609               2.60           2.34                 35    46.7808   
2610               2.60           2.36                 35    46.7808   
2611               2.60           2.38                 34    44.6952   
2612               2.60           2.40                 33    42.5920   
2613               2.60           2.42                 31    40.4712   
2614               2.60           2.44                 31    40.4712   
2615               2.60           2.46                 29    38.3064   
2548               2.55           2.34                 27    35.8688   
2547               2.55           2.32                 27    35.8688   

         yield  money_won_over_all_matches  
2608  1.336594                     11.7808  
2609  1.336594                     11.7808  
2610  1.336594                     11.7808  
2611  1.314565                     10.6952  
2612  1.290667                      9.5920  
2613  1.305523                      9.4712  
2614  1.305523                      9.4712  
2615  1.320910                      9.3064  
2548  1.328474                      8.8688  
2547  1.328474                      8.8688

In [1109]:
under_conditions.to_csv('under_conditions.csv', index= False)

In [98]:
all_considered_matches_under = pd.DataFrame(columns = next_fixtures4.columns)
for index, row in under_conditions.iterrows():
    next_fixtures4['stake'] = next_fixtures4.apply(money_won_goals_under, axis=1,args=(row['goals_total_under'],))
    matches_to_add = next_fixtures4.loc[(next_fixtures4['goals_total_pred']<row['goals_total_under']) & (next_fixtures4['Avg<2.5']>row['under_2_5_val'])]
    all_considered_matches_under = pd.concat([all_considered_matches_under, matches_to_add])
all_considered_matches_under.drop_duplicates(inplace= True)

C:\Users\ciesl\AppData\Local\Temp\ipykernel_36744\3657945737.py:5: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  all_considered_matches_under = pd.concat([all_considered_matches_under, matches_to_add])


In [99]:
(all_considered_matches_under['stake'].sum()*0.88/len(all_considered_matches_under)-1)*len(all_considered_matches_under)

11.780800000000005

In [117]:
under_conditions

goals_total_under  under_2_5_val  number_of_matches  money_won  \
2608               2.60           2.32                 35    46.7808   
2609               2.60           2.34                 35    46.7808   
2610               2.60           2.36                 35    46.7808   
2611               2.60           2.38                 34    44.6952   
2612               2.60           2.40                 33    42.5920   
2613               2.60           2.42                 31    40.4712   
2614               2.60           2.44                 31    40.4712   
2615               2.60           2.46                 29    38.3064   
2548               2.55           2.34                 27    35.8688   
2547               2.55           2.32                 27    35.8688   

         yield  money_won_over_all_matches  
2608  1.336594                     11.7808  
2609  1.336594                     11.7808  
2610  1.336594                     11.7808  
2611  1.314565                     10.6952  
2612  1.290667                      9.5920  
2613  1.305523                      9.4712  
2614  1.305523                      9.4712  
2615  1.320910                      9.3064  
2548  1.328474                      8.8688  
2547  1.328474                      8.8688

### Summary:

- **Over 2.5 goals betting** offers a larger pool of matches to bet on, since games with higher goal totals tend to be more common, especially in certain leagues or when teams with strong offensive capabilities face each other. This provides more opportunities to place bets. However, due to the higher frequency of such matches and the relatively lower odds associated with them, the yield is smaller, meaning that the profit margin per match is reduced. Nonetheless, when we aggregate the bets across all qualifying matches, this strategy still results in a respectable cumulative profit of **4.83 units**. This makes it a steady, lower-risk option with more frequent opportunities to bet.

- **Under 2.5 goals betting**, on the other hand, involves fewer matches because lower-scoring games tend to be less frequent, especially in leagues where attacking play is dominant. As a result, the number of qualifying matches for under 2.5 goals is smaller. However, the **higher yield** in these bets compensates for the lower volume. The odds for under 2.5 goals are often more favorable, and when combined with our model's accuracy in predicting lower-scoring games, this strategy yields a **total profit of over 11 units** across all matches. This makes it a higher-reward option, though it may come with slightly fewer betting opportunities.

- By **combining both strategies**, we can create a more balanced and diversified betting approach. The higher volume of bets in the **over 2.5 goals** market ensures consistent betting opportunities and gradual profit accumulation, while the higher profitability per match in the **under 2.5 goals** market boosts the overall returns. This blend allows us to take advantage of both high-scoring and low-scoring matches, optimizing the potential for profit while spreading the risk. By leveraging the strengths of both strategies, we are better positioned to maximize returns and mitigate the impact of unpredictable match outcomes, ensuring a more stable long-term betting strategy.
